File: internal/domain/transaction/entity/transaction.go
package entity

import (
	"time"

	"github.com/shopspring/decimal"

	"github.com/paxyside/nox-wallet/pkg/ethkit"
)

type Transaction struct {
	ID string
	// UniqueID is Alchemy's "<hash>:<category>:<log_index>" — disambiguates
	// multi-leg events (e.g. a swap = USDC out + OTHER in, same Hash,
	// different UniqueID). Used as the upsert key in storage.
	UniqueID    string
	Hash        string
	From        ethkit.Address
	To          ethkit.Address
	Value       decimal.Decimal // human-readable units
	Asset       string          // "ETH", "USDC", etc.
	Category    ethkit.AssetTransferCategory
	BlockNumber uint64
	Timestamp   time.Time
	CachedAt    time.Time

	// Gas fee fields — populated asynchronously from eth_getTransactionReceipt.
	// Empty string means "not yet fetched".
	GasFeeEth string // e.g. "0.000042"
	GasFeeUsd string // e.g. "$0.11"

	// Swap synthetic fields — populated by the history usecase after grouping
	// two same-hash legs that touch a known DEX router. Persisted rows always
	// have these empty; the merger fills them on read. UI uses IsSwap to pick
	// the swap rendering path.
	IsSwap      bool
	TokenInSym  string // sent token, e.g. "USDC"
	TokenInVal  string // sent amount, e.g. "27.39"
	TokenOutSym string // received token, e.g. "USDT"
	TokenOutVal string // received amount, e.g. "26.91"
}

File: ui/lib/features/history/domain/transaction.dart
/// Domain entity representing a single history transaction.
class Transaction {
  const Transaction({
    required this.txHash,
    required this.from,
    required this.to,
    required this.asset,
    required this.value,
    required this.category,
    required this.blockNum,
    required this.blockTime,
    required this.walletAddress,
    this.gasFeeEth = '',
    this.gasFeeUsd = '',
    this.valueUsd = '',
    this.isSwap = false,
    this.tokenInSym = '',
    this.tokenInValue = '',
    this.tokenOutSym = '',
    this.tokenOutValue = '',
  });

  final String txHash;
  final String from;
  final String to;
  final String asset;
  final String value;
  final String category;
  final int blockNum;
  final DateTime blockTime;

  /// The address of the current wallet — used to derive direction.
  final String walletAddress;

  /// Gas fee in ETH (empty if not yet fetched from receipt).
  final String gasFeeEth;

  /// Gas fee in USD (empty if not yet fetched from receipt).
  final String gasFeeUsd;

  /// USD equivalent of the transaction value (empty if price unavailable).
  final String valueUsd;

  /// True when the server merged two same-hash legs that crossed a known DEX
  /// router into one synthetic swap entry.
  final bool isSwap;

  /// Sent token symbol on a swap (e.g. "USDC"). Empty unless [isSwap].
  final String tokenInSym;

  /// Sent token amount, human-readable. Empty unless [isSwap].
  final String tokenInValue;

  /// Received token symbol on a swap (e.g. "USDT"). Empty unless [isSwap].
  final String tokenOutSym;

  /// Received token amount, human-readable. Empty unless [isSwap].
  final String tokenOutValue;

  /// Returns true when this wallet is the recipient of the transaction.
  bool get isIncoming => to.toLowerCase() == walletAddress.toLowerCase();
}

/// A single page of history items returned from the repository.
class HistoryPage {
  const HistoryPage({
    required this.items,
    required this.nextCursor,
    required this.hasMore,
    this.totalCount = 0,
  });

  final List<Transaction> items;
  final String nextCursor;
  final bool hasMore;

  /// Total number of transactions in the DB for this address.
  final int totalCount;
}
